qiita_db/software.py

```diff
@@ -0,0 +1,522 @@
+            sql_params = [[self.id, doi] for doi, _ in publications]
+            TRN.add(sql, sql_params, many=True)
+            TRN.execute()
+
+
+class Parameters(object):
+    """Models a specific set of parameters of a command
+
+    Attributes
+    ----------
+    name
+    values
+
+    Methods
+    -------
+    exists
+    create
+    iter
+    to_str
+    to_file
+
+    See Also
+    --------
+    qiita_db.software.Command
+    """
+
+    def __eq__(self, other):
+        """Self and other are equal based on type, database id and table"""
+        if type(self) != type(other):
+            return False
+        if other._table != self._table:
+            return False
+        if other.id != self.id:
+            return False
+        return True
+
+    def __init__(self, id_, command):
+        """Initializes the object
+
+        Parameters
+        ----------
+        id_: int
+            The parameter set identifier
+        command : qiita_db.software.Command
+            The command to which the parameter is requested
+
+        Raises
+        ------
+        QiitaDBUnknownIDError
+            If `id_` does not correspond to a parameter set for the given
+            command
+        """
+        with TRN:
+            self._table = command.parameters_table
+            self.id = id_
+            sql = """SELECT EXISTS(
+                        SELECT *
+                        FROM qiita.{0}
+                        WHERE parameters_id = %s)""".format(self._table)
+            TRN.add(sql, [self.id])
+            if not TRN.execute_fetchlast():
+                raise QiitaDBUnknownIDError(self.id, self._table)
+
+    @classmethod
+    def exists(cls, command, **kwargs):
+        r"""Check if a parameter set already exists
+
+        Parameters
+        ----------
+        command : qiita_db.software.Command
+            The command to which the parameter set belongs to
+        kwargs : dict
+            The parameters and their values
+
+        Returns
+        -------
+        bool
+            Whether if the parameter set exists in the given command
+        """
+        with TRN:
+            table = command.parameters_table
+
+            db_cols = set(get_table_cols(table))
+            db_cols.remove("param_set_name")
+            db_cols.remove("parameters_id")
+            missing = db_cols.difference(kwargs)
+
+            if missing:
+                raise QiitaDBError(
+                    "Missing parameters for command %s: %s"
+                    % (command.name, ', '.join(missing)))
+
+            extra = set(kwargs).difference(db_cols)
+            if extra:
+                raise QiitaDBError(
+                    "Extra parameters for command %s: %s"
+                    % (command.name, ', '.join(extra)))
+
+            cols = ["{} = %s".format(col) for col in kwargs]
+            sql = "SELECT EXISTS(SELECT * FROM qiita.{0} WHERE {1})".format(
+                table, ' AND '.join(cols))
```

Comment:
Could there be an instance where the command has no parameters? If so wouldn't such a case make this query be malformed?